### DISCUSSION TITLE
[@types/node] How to automatically install the correct version for the current Node.js environment

### DISCUSSION THREAD

**Discussion by bartfeenstra** [DefinitelyTyped/DefinitelyTyped#69418] [_Misc]
Several discussions within this repo about which `@types/nodes` version to use in projects are answered with "the `@types/node` major version must match your environment's Node.js major versoin". E.g. when running Node.js 20, `@types/node` 20.*.* should be used.

This is easily done for projects that are compatible with a single Node.js version.

For projects that are compatible with a range of major Node.js versions, such as `>= 18`, how would one ensure the correct major version of `@types/node` is installed for the current environment's Node.js version?

**Comment by bartfeenstra** [comment]
Could it be an option to specify engine compatibility in `package.json` for each release?

**Reply by davidenke** [reply]
I encouter this quite often when using `npx -y npm-check-updates -u` as well, as it always uses the latest versions.

Right now for example, Node LTS is at major 22 but 24 is already out. Thus, the latest `@types/node` package is at the time of writing (20.08.2025) `24.3.0`, but I need `22.17.2`. One could just use the semver ranges to achieve this, e.g. setting `@types/node` to `^22` or `22.x.x`. However, as I want to pin my dependencies, this is not an option, so I came up with this:

```bash
# update node and persist version
nvm install --lts && echo $(node -v) > .nvmrc # → v.22.18.0

# update dependencies in manifest
npx -y npm-check-updates -u

# align node types version
NODE_VERSION=$(node -v) # → v.22.18.0
NODE_MAJOR_VERSION=${NODE_VERSION:1:2} # → 22
npm install --save-dev --save-exact @types/node@${NODE_MAJOR_VERSION} # → 22.17.2

# update peer dependencies as well
npm update
```

**Reply by bartfeenstra** [reply]
I've been using an approach like that but it only works in cases where it is okay to make changes to `package.json`.

**Reply by janpaepke** [reply]
For anyone else having this problem: [Here's a guard](https://www.npmjs.com/package/check-node-types) that at least lets you safely disable update checks for @types/node.

**Reply by ghostdevv** [reply]
What about minor ranges? Is it true that if I'm on Node `20.12` (for example), then I want the same `@types/node@20.12.x` rather than `@types/node@20.19.39` (latest 20.x at time of writing)
